woocommerce/woocommerce-gateway-stripe
# Create mandate for subscriptions which do not have a mandate created previously.

Fixes #2607 

The changes in this PR fix a few issues related to mandate in a subscription and renewal. The main target of this PR is to fix the error that happens while processing a `0` amount subscription which needs a mandate. A mandate is required only for the cards issued in India.

The main issues handled in this PR are-

1. Unable to process 0/month subscriptions. We [add](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L598-L612) the `mandate` data to the request and on Stripe end it creates a mandate if required by attaching a payment method to it. The amount needs to be a positive integer and so it fails when the amount is `0`.
2. If we just bypass the mandate options for `0` amount subscriptions and later update the subscription amount to a positive value to be charged, we need to create a mandate and save it for later. 
3. If we change the subscription amount or the payment method, the existing mandate won't work and we need to create a new mandate for the amount and/or the payment method.
4. When UPE is enabled, if we buy a subscription with a saved card it does not create a mandate. 
5. When UPE is disabled, if we buy a subscription with a saved card it creates a mandate and use the same mandate in renewal, But the manual renewal fails.


## Changes proposed in this Pull Request:
 - avoid setting mandate options when the subscription amount is 0. (fixes 1)
 - add mandate data to renewal if not present in any previous renewal orders. (fixes 2)
 - Save mandate id in the order meta, in case it was created in some later renewals, and reuse it on the later renewals. (fixes 2)
 - If the saved mandate is for a different amount and/or different payment method, create a mandate again and save it to the current order meta (fixes 3)
 - Pass `confirm` and `setup_future_usage` attributes correctly when an existing `mandate` is passed with the request. (fixes 4, 5) https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/947acf13125292501761682cb945426d6721d535 https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/b8d987dc8bde561942211c73deea33d7f76b5b64

### Related resources for reading
- https://stripe.com/docs/invoicing/india-emandate-guide
- https://stripe.com/docs/india-recurring-payments?integration=subscriptions

## Testing instructions
### Prerequisite 
- Change the store currency to INR
- Make sure the webhooks are working

### Testing 0 amount subscription (Test with both UPE enabled and disabled)
- Create a subscription with 0 amount and a sign-up fee.
- As a shopper go to the shop page and add the subscription to your cart.
- Use card `4000003560000123` to complete the purchase.
- Make sure that there is no error and the subscription is active.
- Check logs in your [Stripe dashboard](https://dashboard.stripe.com/test/logs?method[0]=post&method[1]=delete&direction[0]=connect_in&direction[1]=self) also to ensure there is no error.
- Find Id of the order, check your DB and make sure that there is no `_stripe_mandate_id` meta set for the order.
- In admin dashboard, go to the subscription edit page and edit the subscription amount to a positive value (500/1000/5000 etc)
- Go to the `my account` page as the shopper and renew the subscription manually.
- Complete the renewal and make sure there is no error.
- Check your DB and make sure that there is now a `_stripe_mandate_id` meta set for the current order.
- Manually renew the order again and check your logs (in Stripe dashboard or in WooCommerce -> Status -> Logs). Make sure that the mandate saved in the previous order is sent with the request.

### Testing saved card with UPE enabled
- Disable UPE
- Create a subscription product with a positive amount
- As a shopper, go to the shop page as a logged in user, make sure you have the `4000003560000123` indian card saved in the payment methods and select the subscription product.
- Complete the purchase
- Make sure a mandate is created. Check your DB and make sure that there is a `_stripe_mandate_id` meta set for this order.
- Manually renew the order again and check your logs (in Stripe dashboard or in WooCommerce -> Status -> Logs). Make sure that the mandate saved in the previous order is sent with the request.

### Testing saved card with UPE disabled
- Enable UPE
- Create a subscription product with a positive amount
- As a shopper, go to the shop page as a logged in user, make sure you have the `4000003560000123` indian card saved in the payment methods and select the subscription product.
- Complete the purchase
- Make sure a mandate is created. Check your DB and make sure that there is a `_stripe_mandate_id` meta set for this order.
- Manually renew the order again and check your logs (in Stripe dashboard or in WooCommerce -> Status -> Logs). Make sure that the mandate saved in the previous order is sent with the request.

### ~~Testing with a different amount~~ (Skip this section)
- ~~Create a subscription with a positive value (500/1000/2000 etc)~~
- ~~As a shopper go to the shop page and add the subscription to your cart.~~
- ~~Use card `4000003560000123` to complete the purchase.~~
- ~~Make sure that there is no error and the subscription is active.~~
- ~~Find Id of the order, check your DB and make sure that there is a `_stripe_mandate_id` meta set for the order.~~
- ~~Renew this order and check the request body in the logs (in Stripe dashboard or in WooCommerce -> Status -> Logs) to ensure that the mandate created in the previous order is sent with the request.~~
- ~~Find Id of the order, check your DB and make sure that there is no `_stripe_mandate_id` meta set for this renewal order.~~
-  ~~In admin dashboard, go to the subscription edit page and edit the subscription amount to a different positive value (1500/3000/5000 etc)~~
- ~~Go to the `my account` page as the shopper and renew the subscription manually.~~
- ~~Find Id of the order, check your DB and make sure that there is a `_stripe_mandate_id` meta set for the order with a different mandate id than before.~~


### Testing critical flows
Make sure all [critical flows](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows) work as intended with both the Classic and UPE checkouts. Make sure to test both Shortcode and Block-based flows. Test with a normal card, 3DS card and Indian card `4000003560000123`

### Testing automatic renewal
Follow and test the steps mentioned in the [original PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2561#issue-1606058756) of implementing e-mandate.

---


**Post merge**
